smart-hadoop-support/smart-hadoop/src/main/java/org/smartdata/hdfs/action/CopyFileAction.java

```diff
@@ -184,42 +187,77 @@ private InputStream getSrcInputStream(String src) throws IOException {
   }
 
   private OutputStream getDestOutPutStream(String dest, long offset) throws IOException {
+    Path destPath = new Path(dest);
+
     if (dest.startsWith("s3")) {
       // Copy to s3
-      FileSystem fs = FileSystem.get(URI.create(dest), getContext().getConf());
-      return fs.create(new Path(dest), true);
+      FileSystem fs = FileSystem.get(destPath.toUri(), getContext().getConf());
+      return fs.create(destPath, true);
     }
 
-    if (dest.startsWith("hdfs")) {
-      // Copy between different clusters
-      // Copy to remote HDFS
-      // Get OutPutStream from URL
-      Configuration remoteClusterConfig = toRemoteClusterConfig(getContext().getConf());
-      FileSystem fs = FileSystem.get(URI.create(dest),  remoteClusterConfig);
-      Path destHdfsPath = new Path(dest);
+    if (PathUtil.isAbsoluteRemotePath(destPath)) {
+      return getRemoteClusterOutputStream(destPath, offset);
+    }
+    return getLocalClusterOutputStream(dest, offset);
+  }
 
-      if (fs.exists(destHdfsPath) && offset != 0) {
-        appendLog("Append to existing file " + dest);
-        return fs.append(destHdfsPath);
+  private OutputStream getLocalClusterOutputStream(String dest, long offset) throws IOException {
+    Optional<FileStatus> destFileStatus = getFileStatusSafely(destPath).map(this::validateDestFile);
+    if (destFileStatus.isPresent() && offset != 0) {
+
+      if (destFileStatus.get().getLen() != offset) {
+        appendLog("Truncating existing local file " + dest + " to the new length " + offset);
+        dfsClient.truncate(dest, offset);
       }
 
-      short replication = getReplication(fs.getDefaultReplication(destHdfsPath));
-      return fs.create(
-          destHdfsPath,
-          true,
-          remoteClusterConfig.getInt(
-              IO_FILE_BUFFER_SIZE_KEY, IO_FILE_BUFFER_SIZE_DEFAULT),
-          replication,
-          fs.getDefaultBlockSize(destHdfsPath));
+      appendLog("Appending to existing local file " + dest);
+      return dfsClient.append(dest, bufferSize, EnumSet.of(CreateFlag.APPEND), null, null);
+    }
+
+    appendLog("Creating new local file " + dest);
+    return dfsClient.create(dest, true,
+        getReplication(dfsClient.getConf().getDefaultReplication()),
+        dfsClient.getConf().getDefaultBlockSize());
+  }
+
+  private FileStatus validateDestFile(FileStatus destFileStatus) {
+    if (destFileStatus.getLen() < offset) {
+      throw new IllegalStateException("Destination file " + destFileStatus.getPath()
```

Comment:
Consider using String.format for readability

Reply:
Fixed